{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","source":["!pip install fastapi nest-asyncio pyngrok uvicorn\n","!pip install tensorflow"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"-XIE6yxGVO48","outputId":"d106ffee-106e-47ec-a404-69b1bd28c77c"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting fastapi\n","  Downloading fastapi-0.111.1-py3-none-any.whl.metadata (26 kB)\n","Requirement already satisfied: nest-asyncio in /usr/local/lib/python3.10/dist-packages (1.6.0)\n","Collecting pyngrok\n","  Downloading pyngrok-7.2.0-py3-none-any.whl.metadata (7.4 kB)\n","Collecting uvicorn\n","  Downloading uvicorn-0.30.3-py3-none-any.whl.metadata (6.5 kB)\n","Collecting starlette<0.38.0,>=0.37.2 (from fastapi)\n","  Downloading starlette-0.37.2-py3-none-any.whl.metadata (5.9 kB)\n","Requirement already satisfied: pydantic!=1.8,!=1.8.1,!=2.0.0,!=2.0.1,!=2.1.0,<3.0.0,>=1.7.4 in /usr/local/lib/python3.10/dist-packages (from fastapi) (2.8.2)\n","Requirement already satisfied: typing-extensions>=4.8.0 in /usr/local/lib/python3.10/dist-packages (from fastapi) (4.12.2)\n","Collecting fastapi-cli>=0.0.2 (from fastapi)\n","  Downloading fastapi_cli-0.0.4-py3-none-any.whl.metadata (7.0 kB)\n","Collecting httpx>=0.23.0 (from fastapi)\n","  Downloading httpx-0.27.0-py3-none-any.whl.metadata (7.2 kB)\n","Requirement already satisfied: jinja2>=2.11.2 in /usr/local/lib/python3.10/dist-packages (from fastapi) (3.1.4)\n","Collecting python-multipart>=0.0.7 (from fastapi)\n","  Downloading python_multipart-0.0.9-py3-none-any.whl.metadata (2.5 kB)\n","Collecting email_validator>=2.0.0 (from fastapi)\n","  Downloading email_validator-2.2.0-py3-none-any.whl.metadata (25 kB)\n","Requirement already satisfied: PyYAML>=5.1 in /usr/local/lib/python3.10/dist-packages (from pyngrok) (6.0.1)\n","Requirement already satisfied: click>=7.0 in /usr/local/lib/python3.10/dist-packages (from uvicorn) (8.1.7)\n","Collecting h11>=0.8 (from uvicorn)\n","  Downloading h11-0.14.0-py3-none-any.whl.metadata (8.2 kB)\n","Collecting dnspython>=2.0.0 (from email_validator>=2.0.0->fastapi)\n","  Downloading dnspython-2.6.1-py3-none-any.whl.metadata (5.8 kB)\n","Requirement already satisfied: idna>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from email_validator>=2.0.0->fastapi) (3.7)\n","Requirement already satisfied: typer>=0.12.3 in /usr/local/lib/python3.10/dist-packages (from fastapi-cli>=0.0.2->fastapi) (0.12.3)\n","Requirement already satisfied: anyio in /usr/local/lib/python3.10/dist-packages (from httpx>=0.23.0->fastapi) (3.7.1)\n","Requirement already satisfied: certifi in /usr/local/lib/python3.10/dist-packages (from httpx>=0.23.0->fastapi) (2024.7.4)\n","Collecting httpcore==1.* (from httpx>=0.23.0->fastapi)\n","  Downloading httpcore-1.0.5-py3-none-any.whl.metadata (20 kB)\n","Requirement already satisfied: sniffio in /usr/local/lib/python3.10/dist-packages (from httpx>=0.23.0->fastapi) (1.3.1)\n","Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from jinja2>=2.11.2->fastapi) (2.1.5)\n","Requirement already satisfied: annotated-types>=0.4.0 in /usr/local/lib/python3.10/dist-packages (from pydantic!=1.8,!=1.8.1,!=2.0.0,!=2.0.1,!=2.1.0,<3.0.0,>=1.7.4->fastapi) (0.7.0)\n","Requirement already satisfied: pydantic-core==2.20.1 in /usr/local/lib/python3.10/dist-packages (from pydantic!=1.8,!=1.8.1,!=2.0.0,!=2.0.1,!=2.1.0,<3.0.0,>=1.7.4->fastapi) (2.20.1)\n","Collecting httptools>=0.5.0 (from uvicorn[standard]>=0.12.0->fastapi)\n","  Downloading httptools-0.6.1-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.6 kB)\n","Collecting python-dotenv>=0.13 (from uvicorn[standard]>=0.12.0->fastapi)\n","  Downloading python_dotenv-1.0.1-py3-none-any.whl.metadata (23 kB)\n","Collecting uvloop!=0.15.0,!=0.15.1,>=0.14.0 (from uvicorn[standard]>=0.12.0->fastapi)\n","  Downloading uvloop-0.19.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (4.9 kB)\n","Collecting watchfiles>=0.13 (from uvicorn[standard]>=0.12.0->fastapi)\n","  Downloading watchfiles-0.22.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (4.9 kB)\n","Collecting websockets>=10.4 (from uvicorn[standard]>=0.12.0->fastapi)\n","  Downloading websockets-12.0-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.6 kB)\n","Requirement already satisfied: exceptiongroup in /usr/local/lib/python3.10/dist-packages (from anyio->httpx>=0.23.0->fastapi) (1.2.2)\n","Requirement already satisfied: shellingham>=1.3.0 in /usr/local/lib/python3.10/dist-packages (from typer>=0.12.3->fastapi-cli>=0.0.2->fastapi) (1.5.4)\n","Requirement already satisfied: rich>=10.11.0 in /usr/local/lib/python3.10/dist-packages (from typer>=0.12.3->fastapi-cli>=0.0.2->fastapi) (13.7.1)\n","Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.10/dist-packages (from rich>=10.11.0->typer>=0.12.3->fastapi-cli>=0.0.2->fastapi) (3.0.0)\n","Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.10/dist-packages (from rich>=10.11.0->typer>=0.12.3->fastapi-cli>=0.0.2->fastapi) (2.16.1)\n","Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.10/dist-packages (from markdown-it-py>=2.2.0->rich>=10.11.0->typer>=0.12.3->fastapi-cli>=0.0.2->fastapi) (0.1.2)\n","Downloading fastapi-0.111.1-py3-none-any.whl (92 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m92.2/92.2 kB\u001b[0m \u001b[31m4.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading pyngrok-7.2.0-py3-none-any.whl (22 kB)\n","Downloading uvicorn-0.30.3-py3-none-any.whl (62 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m62.8/62.8 kB\u001b[0m \u001b[31m5.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading email_validator-2.2.0-py3-none-any.whl (33 kB)\n","Downloading fastapi_cli-0.0.4-py3-none-any.whl (9.5 kB)\n","Downloading h11-0.14.0-py3-none-any.whl (58 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m58.3/58.3 kB\u001b[0m \u001b[31m5.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading httpx-0.27.0-py3-none-any.whl (75 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m75.6/75.6 kB\u001b[0m \u001b[31m7.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading httpcore-1.0.5-py3-none-any.whl (77 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m77.9/77.9 kB\u001b[0m \u001b[31m6.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading python_multipart-0.0.9-py3-none-any.whl (22 kB)\n","Downloading starlette-0.37.2-py3-none-any.whl (71 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m71.9/71.9 kB\u001b[0m \u001b[31m7.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading dnspython-2.6.1-py3-none-any.whl (307 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m307.7/307.7 kB\u001b[0m \u001b[31m10.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading httptools-0.6.1-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (341 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m341.4/341.4 kB\u001b[0m \u001b[31m29.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading python_dotenv-1.0.1-py3-none-any.whl (19 kB)\n","Downloading uvloop-0.19.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.4 MB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.4/3.4 MB\u001b[0m \u001b[31m56.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading watchfiles-0.22.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.2 MB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.2/1.2 MB\u001b[0m \u001b[31m50.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hDownloading websockets-12.0-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (130 kB)\n","\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m130.2/130.2 kB\u001b[0m \u001b[31m13.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hInstalling collected packages: websockets, uvloop, python-multipart, python-dotenv, pyngrok, httptools, h11, dnspython, watchfiles, uvicorn, starlette, httpcore, email_validator, httpx, fastapi-cli, fastapi\n","Successfully installed dnspython-2.6.1 email_validator-2.2.0 fastapi-0.111.1 fastapi-cli-0.0.4 h11-0.14.0 httpcore-1.0.5 httptools-0.6.1 httpx-0.27.0 pyngrok-7.2.0 python-dotenv-1.0.1 python-multipart-0.0.9 starlette-0.37.2 uvicorn-0.30.3 uvloop-0.19.0 watchfiles-0.22.0 websockets-12.0\n","Requirement already satisfied: tensorflow in /usr/local/lib/python3.10/dist-packages (2.15.0)\n","Requirement already satisfied: absl-py>=1.0.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (1.4.0)\n","Requirement already satisfied: astunparse>=1.6.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (1.6.3)\n","Requirement already satisfied: flatbuffers>=23.5.26 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (24.3.25)\n","Requirement already satisfied: gast!=0.5.0,!=0.5.1,!=0.5.2,>=0.2.1 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (0.6.0)\n","Requirement already satisfied: google-pasta>=0.1.1 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (0.2.0)\n","Requirement already satisfied: h5py>=2.9.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (3.11.0)\n","Requirement already satisfied: libclang>=13.0.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (18.1.1)\n","Requirement already satisfied: ml-dtypes~=0.2.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (0.2.0)\n","Requirement already satisfied: numpy<2.0.0,>=1.23.5 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (1.26.4)\n","Requirement already satisfied: opt-einsum>=2.3.2 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (3.3.0)\n","Requirement already satisfied: packaging in /usr/local/lib/python3.10/dist-packages (from tensorflow) (24.1)\n","Requirement already satisfied: protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.20.3 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (3.20.3)\n","Requirement already satisfied: setuptools in /usr/local/lib/python3.10/dist-packages (from tensorflow) (71.0.4)\n","Requirement already satisfied: six>=1.12.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (1.16.0)\n","Requirement already satisfied: termcolor>=1.1.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (2.4.0)\n","Requirement already satisfied: typing-extensions>=3.6.6 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (4.12.2)\n","Requirement already satisfied: wrapt<1.15,>=1.11.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (1.14.1)\n","Requirement already satisfied: tensorflow-io-gcs-filesystem>=0.23.1 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (0.37.1)\n","Requirement already satisfied: grpcio<2.0,>=1.24.3 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (1.64.1)\n","Requirement already satisfied: tensorboard<2.16,>=2.15 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (2.15.2)\n","Requirement already satisfied: tensorflow-estimator<2.16,>=2.15.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (2.15.0)\n","Requirement already satisfied: keras<2.16,>=2.15.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (2.15.0)\n","Requirement already satisfied: wheel<1.0,>=0.23.0 in /usr/local/lib/python3.10/dist-packages (from astunparse>=1.6.0->tensorflow) (0.43.0)\n","Requirement already satisfied: google-auth<3,>=1.6.3 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow) (2.27.0)\n","Requirement already satisfied: google-auth-oauthlib<2,>=0.5 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow) (1.2.1)\n","Requirement already satisfied: markdown>=2.6.8 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow) (3.6)\n","Requirement already satisfied: requests<3,>=2.21.0 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow) (2.31.0)\n","Requirement already satisfied: tensorboard-data-server<0.8.0,>=0.7.0 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow) (0.7.2)\n","Requirement already satisfied: werkzeug>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.16,>=2.15->tensorflow) (3.0.3)\n","Requirement already satisfied: cachetools<6.0,>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from google-auth<3,>=1.6.3->tensorboard<2.16,>=2.15->tensorflow) (5.4.0)\n","Requirement already satisfied: pyasn1-modules>=0.2.1 in /usr/local/lib/python3.10/dist-packages (from google-auth<3,>=1.6.3->tensorboard<2.16,>=2.15->tensorflow) (0.4.0)\n","Requirement already satisfied: rsa<5,>=3.1.4 in /usr/local/lib/python3.10/dist-packages (from google-auth<3,>=1.6.3->tensorboard<2.16,>=2.15->tensorflow) (4.9)\n","Requirement already satisfied: requests-oauthlib>=0.7.0 in /usr/local/lib/python3.10/dist-packages (from google-auth-oauthlib<2,>=0.5->tensorboard<2.16,>=2.15->tensorflow) (1.3.1)\n","Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests<3,>=2.21.0->tensorboard<2.16,>=2.15->tensorflow) (3.3.2)\n","Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests<3,>=2.21.0->tensorboard<2.16,>=2.15->tensorflow) (3.7)\n","Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests<3,>=2.21.0->tensorboard<2.16,>=2.15->tensorflow) (2.0.7)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests<3,>=2.21.0->tensorboard<2.16,>=2.15->tensorflow) (2024.7.4)\n","Requirement already satisfied: MarkupSafe>=2.1.1 in /usr/local/lib/python3.10/dist-packages (from werkzeug>=1.0.1->tensorboard<2.16,>=2.15->tensorflow) (2.1.5)\n","Requirement already satisfied: pyasn1<0.7.0,>=0.4.6 in /usr/local/lib/python3.10/dist-packages (from pyasn1-modules>=0.2.1->google-auth<3,>=1.6.3->tensorboard<2.16,>=2.15->tensorflow) (0.6.0)\n","Requirement already satisfied: oauthlib>=3.0.0 in /usr/local/lib/python3.10/dist-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<2,>=0.5->tensorboard<2.16,>=2.15->tensorflow) (3.2.2)\n"]}]},{"cell_type":"code","execution_count":null,"metadata":{"id":"xdEtVa4bS_hz"},"outputs":[],"source":["# Importar las bibliotecas necesarias\n","from fastapi import FastAPI, File, UploadFile\n","from pydantic import BaseModel\n","from fastapi.middleware.cors import CORSMiddleware\n","from tensorflow.keras.models import load_model\n","from tensorflow.keras.preprocessing import image\n","import cv2\n","import numpy as np\n","import json\n","import base64\n","import io\n","import uvicorn\n","from pyngrok import ngrok\n","import asyncio\n","import nest_asyncio\n","from concurrent.futures import ThreadPoolExecutor"]},{"cell_type":"code","source":["from google.colab import drive\n","# Montar Google Drive\n","drive.mount('/content/drive')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"GlNXLuROUml9","outputId":"290c4b0a-0f7b-4bae-f37d-13619dccbaf0"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/drive\n"]}]},{"cell_type":"code","source":["# Aplicar nest_asyncio para permitir el anidamiento de bucles de eventos\n","nest_asyncio.apply()\n","\n","# Configurar el authtoken de ngrok\n","ngrok.set_auth_token(\"2jqS67zz1ACVJgG3plcESb0ssJO_5P9GqkrZiA4hr9BaeX3oX\")\n","\n","\n","app = FastAPI()\n","\n","# Configurar CORS\n","app.add_middleware(\n","    CORSMiddleware,\n","    allow_origins=[\"*\"],\n","    allow_credentials=True,\n","    allow_methods=[\"*\"],\n","    allow_headers=[\"*\"],\n",")\n","\n","\n","# Configuración\n","IMG_SIZE = 64\n","\n","# Cargar el modelo y las etiquetas\n","model = load_model('/content/drive/MyDrive/Personales/Estudio/BigData/Mod11/TareaFinal/face_recognition_model.h5')\n","\n","def load_labels(filename):\n","    with open(filename, 'r') as f:\n","        return json.load(f)\n","\n","label_mapping = load_labels('/content/drive/MyDrive/Personales/Estudio/BigData/Mod11/TareaFinal/face_recognition_model.json')\n","label_names = {v: k for k, v in label_mapping.items()}\n","\n","\n","def preprocess_image(image, target_size=(64, 64)):\n","    # Convertir a RGB si es necesario\n","    if len(image.shape) == 2 or image.shape[2] == 1:\n","        image = cv2.cvtColor(image, cv2.COLOR_GRAY2RGB)\n","    elif image.shape[2] == 3:\n","        image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)\n","\n","    # Detección facial\n","    face_cascade = cv2.CascadeClassifier(cv2.data.haarcascades + 'haarcascade_frontalface_default.xml')\n","    gray = cv2.cvtColor(image, cv2.COLOR_RGB2GRAY)\n","    faces = face_cascade.detectMultiScale(gray, 1.3, 5)\n","\n","    if len(faces) > 0:\n","        x, y, w, h = faces[0]\n","        face = image[y:y+h, x:x+w]\n","    else:\n","        return None\n","\n","    # Redimensionar\n","    face = cv2.resize(face, target_size)\n","\n","    # Normalizar\n","    face = face.astype(np.float32) / 255.0\n","\n","    return face\n","\n","def identify_person(image):\n","    processed_img = preprocess_image(image, target_size=(IMG_SIZE, IMG_SIZE))\n","\n","    if processed_img is None:\n","        return None, 0\n","\n","    # Preparar la imagen para la predicción\n","    img_array = np.expand_dims(processed_img, axis=0)\n","\n","    # Realizar la predicción\n","    prediction = model.predict(img_array)\n","    person_index = np.argmax(prediction)\n","    confidence = prediction[0][person_index]\n","\n","    # Si la confianza es menor a 0.85, retornar None\n","    if confidence < 0.85:\n","        return None, confidence\n","\n","    person_name = label_names[person_index]\n","    return person_name, confidence\n","\n","class IdentificationResponse(BaseModel):\n","    person: str | None\n","    confidence: float\n","@app.post(\"/identify\", response_model=IdentificationResponse)\n","async def identify_image(file: UploadFile = File(...)):\n","    contents = await file.read()\n","    nparr = np.fromstring(contents, np.uint8)\n","    img = cv2.imdecode(nparr, cv2.IMREAD_COLOR)\n","\n","    person, confidence = identify_person(img)\n","\n","    return IdentificationResponse(person=person, confidence=float(confidence))\n","\n","# Configurar ngrok\n","ngrok_tunnel = ngrok.connect(8004)\n","print('URL pública:', ngrok_tunnel.public_url)\n","\n","# Función para ejecutar el servidor\n","async def run_server():\n","    config = uvicorn.Config(app, host=\"0.0.0.0\", port=8004, log_level=\"info\")\n","    server = uvicorn.Server(config)\n","    await server.serve()\n","\n","# Ejecutar el servidor en el bucle de eventos actual\n","loop = asyncio.get_event_loop()\n","with ThreadPoolExecutor() as executor:\n","    future = executor.submit(loop.run_until_complete, run_server())\n","\n","    try:\n","        future.result()\n","    except KeyboardInterrupt:\n","        print(\"Servidor detenido\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"jJikeuCoUi6s","outputId":"cc93ce15-4ada-4c90-c451-dc8f7c599d02"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["URL pública: https://cb57-34-106-47-40.ngrok-free.app\n"]},{"output_type":"stream","name":"stderr","text":["INFO:     Started server process [181]\n","INFO:     Waiting for application startup.\n","INFO:     Application startup complete.\n","INFO:     Uvicorn running on http://0.0.0.0:8004 (Press CTRL+C to quit)\n","ERROR:asyncio:Exception in callback BaseSelectorEventLoop._accept_connection(<function Ser...x7a381816bf40>, <socket.socke...0.0.0', 8004)>, None, <Server socke....0', 8004)>,)>, 2048, None)\n","handle: <Handle BaseSelectorEventLoop._accept_connection(<function Ser...x7a381816bf40>, <socket.socke...0.0.0', 8004)>, None, <Server socke....0', 8004)>,)>, 2048, None)>\n","Traceback (most recent call last):\n","  File \"/usr/lib/python3.10/asyncio/events.py\", line 80, in _run\n","    self._context.run(self._callback, *self._args)\n","RuntimeError: cannot enter context: <_contextvars.Context object at 0x7a38988a83c0> is already entered\n","ERROR:asyncio:Exception in callback _SelectorSocketTransport._read_ready()\n","handle: <Handle _SelectorSocketTransport._read_ready()>\n","Traceback (most recent call last):\n","  File \"/usr/lib/python3.10/asyncio/events.py\", line 80, in _run\n","    self._context.run(self._callback, *self._args)\n","RuntimeError: cannot enter context: <_contextvars.Context object at 0x7a3818467600> is already entered\n"]},{"output_type":"stream","name":"stdout","text":["INFO:     181.115.186.26:0 - \"GET / HTTP/1.1\" 404 Not Found\n","INFO:     181.115.186.26:0 - \"GET /favicon.ico HTTP/1.1\" 404 Not Found\n","INFO:     181.115.186.26:0 - \"GET /docs HTTP/1.1\" 200 OK\n","INFO:     181.115.186.26:0 - \"GET /openapi.json HTTP/1.1\" 200 OK\n"]},{"output_type":"stream","name":"stderr","text":["ERROR:asyncio:Exception in callback BaseAsyncIOLoop._handle_events(9, 1)\n","handle: <Handle BaseAsyncIOLoop._handle_events(9, 1)>\n","Traceback (most recent call last):\n","  File \"/usr/lib/python3.10/asyncio/events.py\", line 80, in _run\n","    self._context.run(self._callback, *self._args)\n","RuntimeError: cannot enter context: <_contextvars.Context object at 0x7a38ba8a7580> is already entered\n","<ipython-input-9-f7059d6dea18>:87: DeprecationWarning: The binary mode of fromstring is deprecated, as it behaves surprisingly on unicode inputs. Use frombuffer instead\n","  nparr = np.fromstring(contents, np.uint8)\n"]},{"output_type":"stream","name":"stdout","text":["1/1 [==============================] - 0s 247ms/step\n","INFO:     181.115.186.26:0 - \"POST /identify HTTP/1.1\" 200 OK\n"]},{"output_type":"stream","name":"stderr","text":["ERROR:asyncio:Exception in callback BaseAsyncIOLoop._handle_events(9, 1)\n","handle: <Handle BaseAsyncIOLoop._handle_events(9, 1)>\n","Traceback (most recent call last):\n","  File \"/usr/lib/python3.10/asyncio/events.py\", line 80, in _run\n","    self._context.run(self._callback, *self._args)\n","RuntimeError: cannot enter context: <_contextvars.Context object at 0x7a38ba8a7580> is already entered\n"]}]}]}